core/src/main/java/org/mskcc/oncotree/utils/MainTypesUtil.java

```diff
@@ -13,31 +30,14 @@
  */
 public class MainTypesUtil {
 
-    public static List<MainType> getMainTypesByTumorTypes(Map<String, TumorType> tumorTypes) {
+    public static List<MainType> getMainTypesByTumorTypes(Set<TumorType> tumorTypes) {
         Set<MainType> mainTypes = new HashSet<>();
         // skip the root node, "TISSUE". Just add it's children
-        for (String tumorTypeName : tumorTypes.keySet()) {
-            TumorType tumorType = tumorTypes.get(tumorTypeName);
-            Map<String, TumorType> children = tumorType.getChildren();
-            for (String code : children.keySet()) {
-                addMainTypesToSet(children.get(code), mainTypes);
-            }
-        }
-        List<MainType> toReturn = new ArrayList<>();
-        toReturn.addAll(mainTypes);
-        return toReturn;
-    }
-
-    private static void addMainTypesToSet(TumorType tumorType, Set<MainType> mainTypes) {
-        if (tumorType.getMainType() != null) {
-            mainTypes.add(tumorType.getMainType());
-        }
-        Map<String, TumorType> children = tumorType.getChildren();
-        if (children.size() > 0) {
-            for (String code : children.keySet()) {
-                TumorType child = children.get(code);
-                addMainTypesToSet(child, mainTypes);
+        for (TumorType tumorType : tumorTypes) {
+            if (tumorType.getMainType() != null && tumorType.getParent() != null) {
+                mainTypes.add(tumorType.getMainType());
             }
         }
+        return new ArrayList<>(mainTypes);
```

Comment:
oh .. it gets the type parameter from the function return type (diamond operator) .. cool.